sw/airborne/modules/gps/gps_datalink.c

```diff
@@ -148,166 +113,109 @@ static void parse_gps_datalink_small(int16_t heading, uint32_t pos_xyz, uint32_t
+  enu_speed_cm.y = (int32_t)((enu_speed_u >> 10) & 0x7FF); // bits 20-10 speed y in cm/s
+  if (enu_speed_cm.y & 0x400) {
+    enu_speed_cm.y |= 0xFFFFF800;  // sign extend for twos complements
+  }
+  enu_speed_cm.z = (int32_t)((enu_speed_u) & 0x3FF); // bits 9-0 speed z in cm/s
+  if (enu_speed_cm.z & 0x200) {
+    enu_speed_cm.z |= 0xFFFFFC00;  // sign extend for twos complements
+  }
 
-  uint32_t now_ts = get_sys_time_usec();
+  /* Convert the cm/s speed with a resolution to meters/s */
+  enu_speed.x = enu_speed_cm.x / 100.0f * EXTERNAL_POSE_SMALL_SPEED_RES;
+  enu_speed.y = enu_speed_cm.y / 100.0f * EXTERNAL_POSE_SMALL_SPEED_RES;
+  enu_speed.z = enu_speed_cm.z / 100.0f * EXTERNAL_POSE_SMALL_SPEED_RES;
 
-  // if selected, publish magnetometer data
-  #if GPS_DATALINK_USE_MAG
-    send_magnetometer(course, now_ts);
-  #endif
+  /* Convert the heading with the 1e4 fraction to radians */
+  float heading = DL_EXTERNAL_POSE_SMALL_heading(buf) / 1e4;
 
-  // Publish GPS data
-  AbiSendMsgGPS(GPS_DATALINK_ID, now_ts, &gps_datalink);
+  gps_datalink_publish(tow, &enu_pos, &enu_speed, heading);
 }
 
-void gps_datalink_parse_REMOTE_GPS(uint8_t *buf)
+void gps_datalink_parse_REMOTE_GPS_LOCAL(uint8_t *buf)
 {
-  if (DL_REMOTE_GPS_ac_id(buf) != AC_ID) { return; } // not for this aircraft
-
-  parse_gps_datalink(DL_REMOTE_GPS_numsv(buf),
-                     DL_REMOTE_GPS_ecef_x(buf),
-                     DL_REMOTE_GPS_ecef_y(buf),
-                     DL_REMOTE_GPS_ecef_z(buf),
-                     DL_REMOTE_GPS_lat(buf),
-                     DL_REMOTE_GPS_lon(buf),
-                     DL_REMOTE_GPS_alt(buf),
-                     DL_REMOTE_GPS_hmsl(buf),
-                     DL_REMOTE_GPS_ecef_xd(buf),
-                     DL_REMOTE_GPS_ecef_yd(buf),
-                     DL_REMOTE_GPS_ecef_zd(buf),
-                     DL_REMOTE_GPS_tow(buf),
-                     DL_REMOTE_GPS_course(buf));
-}
+  if (DL_REMOTE_GPS_LOCAL_ac_id(buf) != AC_ID) { return; } // not for this aircraft
 
-void gps_datalink_parse_REMOTE_GPS_SMALL(uint8_t *buf)
-{
-  if (DL_REMOTE_GPS_SMALL_ac_id(buf) != AC_ID) { return; } // not for this aircraft
+  uint32_t tow = DL_REMOTE_GPS_LOCAL_tow(buf);
 
-  parse_gps_datalink_small(DL_REMOTE_GPS_SMALL_heading(buf),
-                           DL_REMOTE_GPS_SMALL_pos_xyz(buf),
-                           DL_REMOTE_GPS_SMALL_speed_xyz(buf),
-                           DL_REMOTE_GPS_SMALL_tow(buf));
-}
+  struct EnuCoor_f enu_pos, enu_speed;
+  enu_pos.x = DL_REMOTE_GPS_LOCAL_enu_x(buf);
+  enu_pos.y = DL_REMOTE_GPS_LOCAL_enu_y(buf);
+  enu_pos.z = DL_REMOTE_GPS_LOCAL_enu_z(buf);
 
-void gps_datalink_parse_REMOTE_GPS_LOCAL(uint8_t *buf)
-{
-  if (DL_REMOTE_GPS_LOCAL_ac_id(buf) != AC_ID) { return; } // not for this aircraft
+  enu_speed.x = DL_REMOTE_GPS_LOCAL_enu_xd(buf);
+  enu_speed.y = DL_REMOTE_GPS_LOCAL_enu_yd(buf);
+  enu_speed.z = DL_REMOTE_GPS_LOCAL_enu_zd(buf);
+
+  float course = DL_REMOTE_GPS_LOCAL_course(buf);
```

Comment:
course is received in deg, should be converted to rad
```suggestion
  float course = RadOfDeg(DL_REMOTE_GPS_LOCAL_course(buf));
```

Reply:
Tnx for this find!